File: 1/binary_gap.java
import java.util.Arrays;
import java.util.Collections;
class Solution {
    public static void main(String args[]) {
        if(args.length==0) {
            System.out.println("input: 1041; output: " + solution(1041));
            System.out.println("input: 15; output: " + solution(15));
            System.out.println("input: 32; output: " + solution(32));
        }
        else
        {
            for(int i = 0; i < args.length; i++) {
                System.out.println("input: " + args[i] + "; output: " + solution(Integer.parseInt(args[i])));
            }
        }
    }
    static int solution (int N) {
        // convert to binary string
        String bin = Integer.toBinaryString(N);
        // remove trailing/leading zeros; split by 1s
        bin = bin.replaceAll("^0+|0+$","");

        // split by 1s
        String[] binArray = bin.split("1");

        // convert each binary gap to a character count
        int gaps[] = new int[binArray.length];
        for(int i = 0; i < binArray.length; i++) {
            gaps[i] = binArray[i].length();
        }

        // obtain maximum number from character counts
        int result = 0;
        for(int i = 0; i < gaps.length; i++) {
            if (gaps[i] > result) {
                result = gaps[i];
            }
        }
        return result;
    }
}

File: 1/binary_gap.pl
#!/usr/bin/perl
use warnings;
use strict;
use Data::Dumper qw( Dumper );
use List::Util qw( max );

sub binary_gap {
    my($N)=@_;

    # convert to binary string
    my $str = unpack("B32", pack("N", shift));

    # remove leading & trailing zeros
    $str =~ s/^0+|0+$//g;

    # split by 1s
    my @arr = split /1/, $str;

    # if there are no 1s, return 0
    if (!@arr)
    {
        return 0;
    }

    # convert each binary gap to a character count
    for my $i (0 .. $#arr)
    {
        $arr[$i] = length ($arr[$i]);
    }
    return max @arr;
}
my @test_array = (9,529,20,15);

foreach my $number(@test_array) {
    my $sol = binary_gap($number);
    print("binary_gap($number): $sol\n");
}
